pkg/apps/watch-appstream.py

```diff
@@ -0,0 +1,383 @@
+        'description': convert_description(element(xml, 'description')),
+        'icon': find_and_convert_icon(dir, origin, xml),
+        'screenshots': convert_screenshots(element(xml, 'screenshots')),
+        'launchables': launchables
+    }
+
+def convert_upstream_component(file, xml):
+    if xml.tag != 'component':
+        return None
+
+    launchables = convert_launchables(xml)
+    if len(launchables) ==  0:
+        return None
+
+    return {
+        'id': element_value(xml, 'id'),
+        'name': element_value(xml, 'name'),
+        'summary': element_value(xml, 'summary'),
+        'description': convert_description(element(xml, 'description')),
+        'icon': find_and_convert_icon(dir, '', xml),
+        'screenshots': convert_screenshots(element(xml, 'screenshots')),
+        'launchables': launchables,
+        'installed': True,
+        'file': file
+    }
+
+class MetainfoDB:
+    def __init__(self):
+        self.dumping = False
+        self.installed_by_file = { }
+        self.available_by_file = { }
+
+    def notice_installed(self, file, xml_root):
+        if xml_root is not None:
+            comp = convert_upstream_component(file, xml_root)
+            if comp is not None:
+                self.installed_by_file[file] = comp;
+        elif file in self.installed_by_file:
+            del self.installed_by_file[file];
+        if self.dumping:
+            self.dump()
+
+    def notice_available(self, file, xml_root):
+        if xml_root is not None:
+            info = { }
+            origin = xml_root.attrib['origin']
+            for xml_comp in xml_root.iter('component'):
+                comp = convert_collection_component(os.path.dirname(file), origin, xml_comp)
+                if comp is not None:
+                    if comp['id'] in info:
+                        pass # warning: duplicate id
+                    else:
+                        info[comp['id']] = comp
+            self.available_by_file[file] = info
+        elif file in self.available_by_file:
+            del self.available_by_file[file]
+        if self.dumping:
+            self.dump()
+
+    def dump(self):
+        comps = { }
+        for file in self.installed_by_file:
+            comp = self.installed_by_file[file]
+            if comp['id'] in comps:
+                pass # warn dup
+            else:
+                comps[comp['id']] = comp;
+        for file in self.available_by_file:
+            for id in self.available_by_file[file]:
+                comp = self.available_by_file[file][id]
+                if not comp['id'] in comps:
+                    comps[comp['id']] = comp;
+                else:
+                    z = comps[comp['id']].copy()
+                    z.update(comp)
+                    comps[comp['id']] = z;
+
+        data = {
+            'components': comps,
+            'origin_files': list(self.available_by_file.keys())
+        }
+
+        print(data)
```

Comment:
Debugging leftover? This would interfere with actual operations, as that also goes to stdout, no?

Reply:
Yes, good catch!